### PR TITLE
feat: support oauth authentication [HEAD-114]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
### Description

Some environments need oauth authentication instead of the Snyk token. This PR enables support authenticating and scanning using an oauth access token.

🚨 Dependent on: https://github.com/snyk/snyk-ls/pull/294

- [x] update token validation to allow json
- [x] spawn cli with different env variables when using oauth
- [x] move snyk learn api calls to language server 
- [x] remove snyk api client and move /user/me call to language server 

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] CHANGELOG.md updated
- [ ] README.md updated, if user-facing

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: src/snyk/common/services/learnService.ts [hunk 1]
```diff
@@ -1,146 +1,60 @@
-import axios from 'axios';
-import { isCodeIssue, isCodeIssueOld, isOssIssue, OpenCommandIssueType } from '../../common/commands/types';
-import { SNYK_LEARN_API_CACHE_DURATION_IN_MS } from '../../common/constants/general';
-import type { completeFileSuggestionType } from '../../snykCode/interfaces';
+import vscode from 'vscode';
```
Comment (michelkaporin) [inline src/snyk/common/services/learnService.ts:1]: You won't be able to unit test this file with this import, but would be good to have tests in place.

Comment (michelkaporin) [reply src/snyk/common/services/learnService.ts:1]: There's `IVSCodeCommands` inversion for command execution.

File: package.json [hunk 2]
```diff
@@ -272,17 +272,17 @@
         {
           "id": "snyk.views.analysis.code.quality",
           "name": "Code Quality",
-          "when": "snyk:lsCodePreview && snyk:loggedIn && snyk:featuresSelected && snyk:codeEnabled && !snyk:codeLocalEngineEnabled && snyk:workspaceFound && !snyk:error"
```
Comment (michelkaporin) [inline package.json:275]: We should keep !codeLocalEngineEnabled context here and above, as we don't want to show results view, when LCE is enabled

File: src/snyk/common/constants/commands.ts [hunk 3]
```diff
@@ -17,6 +17,8 @@ export const SNYK_OPEN_ISSUE_COMMAND = 'snyk.showissue';
 export const SNYK_IGNORE_ISSUE_COMMAND = 'snyk.ignoreissue';
 export const SNYK_SHOW_OUTPUT_COMMAND = 'snyk.showOutputChannel';
 export const SNYK_SHOW_LS_OUTPUT_COMMAND = 'snyk.showLsOutputChannel';
+export const SNYK_GET_LESSON_COMMAND = 'snyk.getLearnLesson';
+export const SNYK_CLI_CONFIG_SAST_ENABLED = 'snyk.cliConfigSettings.sastEnabled';
```
Comment (michelkaporin) [inline src/snyk/common/constants/commands.ts:21]: Why not to name command consistently with others in LS as a verb, e.g. `get...`?

File: src/snyk/snykCode/codeServiceOld.ts [hunk 4]
```diff
@@ -190,7 +187,7 @@ export class SnykCodeServiceOld extends AnalysisStatusProvider implements ISnykC
         },
       };
 
-      let result: FileAnalysis | null = null;
+      let result: FileAnalysis | null;
```
[the remark this answers is not in the document]
Comment (bastiandoetsch) [reply src/snyk/snykCode/codeServiceOld.ts:190]: webstorm told me it's unnecessary ;)

File: .runUnit/Unit Tests.run.xml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<component name="ProjectRunConfigurationManager">
```
Comment (michelkaporin) [inline .runUnit/Unit Tests.run.xml:1]: Is it an IntelliJ tests run definition? Haven't seen such folder before.

File: src/snyk/common/services/learnService.ts [hunk 6]
```diff
@@ -1,146 +1,62 @@
...
+    if (additionalData.cwe.length > 0) {
+      cwe = additionalData.cwe[0];
+    }
 
-        params = LearnService.getOSSIssueParams(issue);
-      } else {
-        ErrorHandler.handle(new Error(`Issue type "${issueType}" not supported`), this.logger);
-        return null;
-      }
+    return this.commandExecutor.executeCommand('snyk.getLearnLesson', rule, ecosystem, cwe, '', 2);
```
Comment (michelkaporin) [inline src/snyk/common/services/learnService.ts:47]: Const exists

**Review by michelkaporin — APPROVED** [review]

File: .runUnit/Unit Tests.run.xml (hunk 5 above)
Comment (bastiandoetsch) [reply .runUnit/Unit Tests.run.xml:1]: yep :)

File: src/snyk/common/services/learnService.ts [hunk 7]
```diff
@@ -1,146 +1,62 @@
...
+    }
 
-      return lessons;
+    const cve = vulnerability.identifiers?.CWE;
+    let cveElement = '';
+    if (cve && cve.length > 0) {
+      cveElement = cve[0];
     }
+    return this.commandExecutor.executeCommand(
+      'snyk.getLearnLesson',
```
[the remark this answers is not in the document]
Comment (bastiandoetsch) [reply src/snyk/common/services/learnService.ts:28]: Oops.

File: src/snyk/snykCode/codeSettings.ts [hunk 8]
```diff
@@ -38,27 +38,37 @@ export class CodeSettings implements ICodeSettings {
...
+    try {
+      const settings = await this.getSastSettings();
+      if (!settings) {
+        return false;
+      }
+      codeEnabled = settings.sastEnabled && !settings.localCodeEngine.enabled;
+      localCodeEngineEnabled = settings.localCodeEngine.enabled;
+    } catch (e) {
+      // Ignore potential command not found error during LS startup and poll
+      codeEnabled = await this.enable(false);
```
Comment (asaf92) [inline src/snyk/snykCode/codeSettings.ts:52]: Why do we try to enable Code here? If it's unavailable to the user, there will be 65+ seconds of polling before we get `false`

**Pull request merged**